libkbfs/rekey_fsm.go

```diff
@@ -0,0 +1,488 @@
+		context.Background(), CtxRekeyIDKey, CtxRekeyOpID, nil))
+}
+
+func newRekeyFinishedEvent(res RekeyResult, err error) RekeyEvent {
+	return RekeyEvent{
+		eventType: rekeyFinishedEvent,
+		finished: &rekeyFinished{
+			RekeyResult: res,
+			err:         err,
+		},
+	}
+}
+
+func newRekeyTimeupEvent() RekeyEvent {
+	return RekeyEvent{
+		eventType: rekeyTimeupEvent,
+	}
+}
+
+func newRekeyShutdownEvent() RekeyEvent {
+	return RekeyEvent{
+		eventType: rekeyShutdownEvent,
+	}
+}
+
+func newRekeyKickoffEventForTest() RekeyEvent {
+	return RekeyEvent{
+		eventType: rekeyKickoffEventForTest,
+	}
+}
+
+func newRekeyCancelEventForTest() RekeyEvent {
+	return RekeyEvent{
+		eventType: rekeyCancelEventForTest,
+	}
+}
+
+// rekeyState models a state in the FSM. rekeyFSM keeps exactly one instance of
+// rekeyState at any given time.
+type rekeyState interface {
+	// reactToEvent defines how this state reacts to an event. Implementations of
+	// rekeyState should handle necessary transition actions in reactToEvent(),
+	// and retur a new rekeyState instance after transition is finished. rekeyFSM
+	// sends event to the rekeyState instance it holds whenever it receives an
+	// event, and use the returned rekeyState instance as new state. It's OK to
+	// return the receiver itself as "new" state.
+	//
+	// rekeyFSM runs an event loop in a dedicated goroutine that calls
+	// reactToEvent and updates states. In other words, it's safe to assume
+	// reactToEvent is only called within the same goroutine, and that it's
+	// impossible that multiple reactToEvent calls are issued concurrently.
+	reactToEvent(event RekeyEvent) rekeyState
+}
+
+type rekeyStateIdle struct {
+	fsm *rekeyFSM
+}
+
+func newRekeyStateIdle(fsm *rekeyFSM) *rekeyStateIdle {
+	return &rekeyStateIdle{fsm: fsm}
+}
+
+func (r *rekeyStateIdle) reactToEvent(event RekeyEvent) rekeyState {
+	switch event.eventType {
+	case rekeyRequestEvent:
+		return newRekeyStateScheduled(r.fsm,
+			event.request.delay, event.request.rekeyTask)
+	default:
+		return r
+	}
+}
+
+type rekeyStateScheduled struct {
+	fsm *rekeyFSM
+
+	timer    *time.Timer
+	deadline time.Time
+
+	task rekeyTask
+}
+
+func newRekeyStateScheduled(
+	fsm *rekeyFSM, delay time.Duration, task rekeyTask) *rekeyStateScheduled {
+	return &rekeyStateScheduled{
+		fsm: fsm,
+		timer: time.AfterFunc(delay, func() {
+			fsm.Event(newRekeyTimeupEvent())
+		}),
+		deadline: time.Now().Add(delay),
+		task:     task,
+	}
+}
+
+func (r *rekeyStateScheduled) reactToEvent(event RekeyEvent) rekeyState {
+	switch event.eventType {
+	case rekeyTimeupEvent:
+		return newRekeyStateStarted(r.fsm, r.task)
+	case rekeyRequestEvent:
+		task := r.task
+		task.promptPaper = task.promptPaper || event.request.promptPaper
+		if task.timeout == nil {
+			task.timeout = event.request.timeout
+		}
+		task.ttl = event.request.ttl
+		r.fsm.log.CDebugf(task.ctx, "Replacing context")
+		task.ctx = event.request.ctx
```

Comment:
This is specifically to handle the stallers, which put staller key in the context. This is actually another reason why the context needs to be persisted through different states.